tasks/seal/task_synth_proofs.go

```diff
@@ -0,0 +1,239 @@
+
+	"github.com/ipfs/go-cid"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-commp-utils/zerocomm"
+	"github.com/filecoin-project/go-state-types/abi"
+
+	"github.com/filecoin-project/curio/harmony/harmonydb"
+	"github.com/filecoin-project/curio/harmony/harmonytask"
+	"github.com/filecoin-project/curio/harmony/resources"
+	"github.com/filecoin-project/curio/lib/ffi"
+	"github.com/filecoin-project/curio/lib/filler"
+	"github.com/filecoin-project/curio/lib/paths"
+
+	"github.com/filecoin-project/lotus/storage/sealer/ffiwrapper"
+	"github.com/filecoin-project/lotus/storage/sealer/storiface"
+)
+
+type SyntheticProofTask struct {
+	sp *SealPoller
+	db *harmonydb.DB
+	sc *ffi.SealCalls
+
+	//max int
+}
+
+func NewSyntheticProofTask(sp *SealPoller, db *harmonydb.DB, sc *ffi.SealCalls) *SyntheticProofTask {
+	return &SyntheticProofTask{
+		sp: sp,
+		db: db,
+		sc: sc,
+
+		//max: maxTrees,
+	}
+}
+
+func (s *SyntheticProofTask) Do(taskID harmonytask.TaskID, stillOwned func() bool) (done bool, err error) {
+	ctx := context.Background()
+
+	var sectorParamsArr []struct {
+		SpID         int64                   `db:"sp_id"`
+		SectorNumber int64                   `db:"sector_number"`
+		RegSealProof abi.RegisteredSealProof `db:"reg_seal_proof"`
+		SealedCID    string                  `db:"tree_r_cid"`
+		UnsealedCID  string                  `db:"tree_d_cid"`
+		TicketValue  []byte                  `db:"ticket_value"`
+	}
+
+	err = s.db.Select(ctx, &sectorParamsArr, `
+		SELECT sp_id, sector_number, reg_seal_proof, tree_d_cid, tree_r_cid, ticket_value
+		FROM sectors_sdr_pipeline
+		WHERE task_id_synth = $1`, taskID)
+	if err != nil {
+		return false, xerrors.Errorf("getting sector params: %w", err)
+	}
+
+	if len(sectorParamsArr) != 1 {
+		return false, xerrors.Errorf("expected 1 sector params, got %d", len(sectorParamsArr))
+	}
+	sectorParams := sectorParamsArr[0]
+
+	// Exit here successfully if synthetic proofs are not required
+	_, ok := abi.Synthetic[sectorParams.RegSealProof]
+	if !ok {
+		serr := s.markFinished(ctx, sectorParams.SpID, sectorParams.SectorNumber)
+		if serr != nil {
+			return false, serr
+		}
+		log.Infof("Not generating synthetic proofs for miner %d sector %d", sectorParams.SpID, sectorParams.SectorNumber)
+		return true, nil
+	}
+
+	sealed, err := cid.Parse(sectorParams.SealedCID)
+	if err != nil {
+		return false, xerrors.Errorf("failed to parse sealed cid: %w", err)
+	}
+
+	unsealed, err := cid.Parse(sectorParams.UnsealedCID)
+	if err != nil {
+		return false, xerrors.Errorf("failed to parse unsealed cid: %w", err)
+	}
+
+	sref := storiface.SectorRef{
+		ID: abi.SectorID{
+			Miner:  abi.ActorID(sectorParams.SpID),
+			Number: abi.SectorNumber(sectorParams.SectorNumber),
+		},
+		ProofType: sectorParams.RegSealProof,
+	}
+
+	var pieces []struct {
+		PieceIndex  int64  `db:"piece_index"`
+		PieceCID    string `db:"piece_cid"`
+		PieceSize   int64  `db:"piece_size"`
+		DataRawSize *int64 `db:"data_raw_size"`
```

Comment:
You definitely don't want raw data size here, just piece size

Reply:
We need that to calculate offset and padding for filler pieces.

Reply:
I am not sure sure if `abi.PaddedPieceSize(PieceSize).Unpadded == DataRawSize` is true.

Reply:
No, DataRawSize is the .car size, without the null-padding making it power-of-two-after-fr32-padding.

You just want to use PieceSize

Reply:
Resolved with new dealData function usage